demos/programs/draw/draw.man
.\" $XConsortium: draw.man /main/4 1995/07/17 10:46:58 drk $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY
.TH draw 1X MOTIF "Demonstration programs"
.SH NAME
\*Ldraw\*O\ - a ScrolledWindow/DrawingArea graphic demo.
.SH SYNOPSIS
.sS
\*Ldraw\*O
.sE
.SH DESCRIPTION
\*LDraw\*O
provides a demonstration of using a DrawingArea in an AUTOMATIC
ScrolledWindow. 

.SH Files
.nf
.ta 1.2i
XmdDraw: the app-default file is very important for usability.
Try running something like: (XENVIRONMENT=XmdDraw draw) 
.fi
.SH X Defaults
The class of this application is \fIXmdDraw\fR.  

.SH Editres: 
Draw is Editres aware, so you can use 'editres' to see the
names of the widgets in the instance hierarchy.



.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


lib/Xm/Draw.c
/*
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
*/
#ifdef REV_INFO
#ifndef lint
static char rcsid[] = "$XConsortium: Draw.c /main/12 1995/10/25 20:02:15 cde-sun $"
#endif
#endif


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif


#include "XmI.h"
#include <Xm/DrawP.h>


/*********************************************************************
 *    Only goes in this module the Draw API used by everybody in Xm.
 *    Anything specific to a class should go in its own Dr*.c module.
 *********************************************************************/

/********    Static Function Declarations    ********/

static void DrawSimpleShadow(Display *display,
			     Drawable d,
			     GC top_gc,
			     GC bottom_gc,
			     Position x,
			     Position y,
			     Dimension width,
			     Dimension height,
			     Dimension shadow_thick,
			     Dimension cor);
/********    End Static Function Declarations    ********/


/*-------------------- Private functions ----------------------*/
/*-------------------------------------------------------------*/

static void
DrawSimpleShadow (Display *display,
		  Drawable d,
		  GC top_gc,
		  GC bottom_gc,
		  Position x,
		  Position y,
		  Dimension width,
		  Dimension height,
		  Dimension shadow_thick,
		  Dimension cor)
/* New implementation (1.2 vs 1.1) uses XSegments instead of XRectangles. */
/* Used for the simple shadow, the etched shadow and the separators */
/* Segment has been faster than Rectangles in all my benches, either
   on Hp, Sun or Pmax. Lines has been slower, that I don't understand... */
{
  static XSegment * segms = NULL;
  static int segm_count = 0;

  register int i, size2, size3;

  if (!d) return;
  ASSIGN_MIN(shadow_thick, (width >> 1));
  ASSIGN_MIN(shadow_thick, (height >> 1));
  if (shadow_thick <= 0) return;

  size2 = (shadow_thick << 1);
  size3 = size2 + shadow_thick;

  _XmProcessLock();
  if (segm_count < shadow_thick) {
    segms = (XSegment *) XtRealloc((char*)segms,
				   sizeof (XSegment) * (size2 << 1));
    segm_count = shadow_thick;
  }

  for (i = 0; i < shadow_thick; i++) {
    /*  Top segments  */
    segms[i].x1 = x;
    segms[i].y2 = segms[i].y1 = y + i;
    segms[i].x2 = x + width - i - 1;
    /*  Left segments  */
    segms[i + shadow_thick].x2 = segms[i + shadow_thick].x1 = x + i;
    segms[i + shadow_thick].y1 = y + shadow_thick;
    segms[i + shadow_thick].y2 = y + height - i - 1;

    /*  Bottom segments  */
    segms[i + size2].x1 = x + i + ((cor)?0:1) ;
    segms[i + size2].y2 = segms[i + size2].y1 = y + height - i - 1;
    segms[i + size2].x2 = x + width - 1 ;
    /*  Right segments  */
    segms[i + size3].x2 = segms[i + size3].x1 = x + width - i - 1;
    segms[i + size3].y1 = y + i + 1 - cor;
    segms[i + size3].y2 = y + height - 1 ;
  }

  XDrawSegments (display, d, top_gc, &segms[0], size2);
  XDrawSegments (display, d, bottom_gc, &segms[size2], size2);
  _XmProcessUnlock();
}

/**************************** Public functions *************************/
/***********************************************************************/

/****************************XmeDrawShadows****************************/
void XmeDrawShadows(Display *display, Drawable d,
                  GC top_gc, GC bottom_gc,
                                           Position x, Position y,
                  Dimension width, Dimension height, Dimension shad_thick,
                  unsigned int shad_type)
{
    GC  tmp_gc ;
    XtAppContext app;

    if(!d) return ;

    app = XtDisplayToApplicationContext(display);

    _XmAppLock(app);

    if ((shad_type == XmSHADOW_IN) || (shad_type == XmSHADOW_ETCHED_IN)) {
        tmp_gc = top_gc ;
        top_gc = bottom_gc ;  /* switch top and bottom shadows */
        bottom_gc = tmp_gc ;
    }

    if ((shad_type == XmSHADOW_ETCHED_IN ||
         shad_type == XmSHADOW_ETCHED_OUT) && (shad_thick != 1)) {
        DrawSimpleShadow (display, d, top_gc, bottom_gc, x, y,
                            width, height, shad_thick/2, 1);
        DrawSimpleShadow (display, d, bottom_gc, top_gc,
                            x + shad_thick/2, y + shad_thick/2,
                            width - (shad_thick/2)*2,
                            height - (shad_thick/2)*2, shad_thick/2, 1);
    } else
        DrawSimpleShadow (display, d, top_gc, bottom_gc, x, y,
                            width, height, shad_thick, 0);
    _XmAppUnlock(app);
}


/*****************************XmeClearBorder*********************************/
void XmeClearBorder (Display *display, Window w,
                                                 Position x, Position y,
                    Dimension width, Dimension height, Dimension shadow_thick)
{
    XtAppContext app;

    if (!w || !shadow_thick || !width || !height) return ;

    app = XtDisplayToApplicationContext(display);
    _XmAppLock(app);

    XClearArea (display, w, x, y, width, shadow_thick, FALSE);
    XClearArea (display, w, x, y + height - shadow_thick, width,
                shadow_thick, FALSE);
    XClearArea (display, w, x, y, shadow_thick, height, FALSE);
    XClearArea (display, w, x + width - shadow_thick, y, shadow_thick,
                height, FALSE);
    _XmAppUnlock(app);
}





/****************************XmeDrawHighlight*************************/
void XmeDrawHighlight(Display *display, Drawable d,
			    GC gc,
                            Position x, Position y,
                            Dimension width, Dimension height,
                            Dimension highlight_thickness)
{
    XRectangle rect[4] ;
    XtAppContext app;

    if (!d || !highlight_thickness || !width || !height) return ;

    app = XtDisplayToApplicationContext(display);

    _XmAppLock(app);

    rect[0].x = rect[1].x = rect[2].x = x ;
    rect[3].x = x + width - highlight_thickness ;
    rect[0].y = rect[2].y = rect[3].y = y ;
    rect[1].y = y + height - highlight_thickness ;
    rect[0].width = rect[1].width = width ;
    rect[2].width = rect[3].width = highlight_thickness ;
    rect[0].height = rect[1].height = highlight_thickness ;
    rect[2].height = rect[3].height = height ;

    XFillRectangles (display, d, gc, rect, 4);

    _XmAppUnlock(app);

}



/******************************XmeDrawSeparator**********************/
void XmeDrawSeparator(Display *display, Drawable d,
                     GC top_gc, GC bottom_gc, GC separator_gc,
                     Position x, Position y,
                     Dimension width, Dimension height,
                     Dimension shadow_thick,
                     Dimension margin, unsigned char orientation,
                     unsigned char separator_type)
{
   Position center;
   XSegment segs[2];
   GC   tmp_gc;
   int i, ndash, shadow_dash_size ;
   XtAppContext app;

   if (!d || (separator_type == XmNO_LINE)) return ;

   app = XtDisplayToApplicationContext(display);
   _XmAppLock(app);

   if (orientation == XmHORIZONTAL) {
       center = y + height / 2;
   } else {
       center = x + width / 2;
   }

   if (separator_type == XmSINGLE_LINE ||
       separator_type == XmSINGLE_DASHED_LINE) {
       if (orientation == XmHORIZONTAL) {
           segs[0].x1 = x + margin;
           segs[0].y1 = segs[0].y2 = center;
           segs[0].x2 = x + width - margin - 1;
       } else {
           segs[0].y1 = y + margin;
           segs[0].x1 = segs[0].x2 = center;
           segs[0].y2 = y + height - margin - 1;
       }
       XDrawSegments (display, d, separator_gc, segs, 1);
       _XmAppUnlock(app);
       return;
   }

   if (separator_type == XmDOUBLE_LINE ||
       separator_type == XmDOUBLE_DASHED_LINE) {
       if (orientation == XmHORIZONTAL) {
           segs[0].x1 = segs[1].x1 = x + margin;
           segs[0].x2 = segs[1].x2 = x + width - margin - 1;
           segs[0].y1 = segs[0].y2 = center - 1;
           segs[1].y1 = segs[1].y2 = center + 1;
       } else {
           segs[0].y1 = segs[1].y1 = y + margin;
           segs[0].y2 = segs[1].y2 = y + height - margin - 1;
           segs[0].x1 = segs[0].x2 = center - 1;
           segs[1].x1 = segs[1].x2 = center + 1;
       }
       XDrawSegments (display, d, separator_gc, segs, 2);
       _XmAppUnlock(app);
       return;
   }

   /* only shadowed stuff in the following, so shadowThickness has to be
      something real */
   if (!shadow_thick) { _XmAppUnlock(app); return ; }

   /* do the in/out effect now */
   if (separator_type == XmSHADOW_ETCHED_IN ||
       separator_type == XmSHADOW_ETCHED_IN_DASH) {
       tmp_gc = top_gc ;
       top_gc = bottom_gc ;
       bottom_gc = tmp_gc ;
   }

   /* In the following, we need to special case the shadow_thick = 2 or 3 case,
      since : it's the default and we don't like changes in visual here,
      and also it looks non symetrical the way it is without special code:
                                    ......
	                            .,,,,,

	and you really want to have ......
                                    ,,,,,,
      So we won't use DrawSimpleShadow in this case, painful but hey..
   */


   /* Now the regular shadowed cases, in one pass with one looong dash
      for the non dashed case */

   if (separator_type == XmSHADOW_ETCHED_IN_DASH ||
       separator_type == XmSHADOW_ETCHED_OUT_DASH)
   /* for now, shadowed dash use three time the shadow thickness as a
      dash size, and worried about the shadow_thick odd values as well */
       shadow_dash_size = (shadow_thick/2)*2*3 ;
   else

   /* regular case, only one dash, the length of the separator */
       shadow_dash_size = (orientation == XmHORIZONTAL)?
           (width - 2*margin):(height - 2*margin) ;

   /* special case for shadowThickness = 1 */
   if (shadow_dash_size == 0) shadow_dash_size = 5 ;

   /* ndash value will be 1 for the regular shadow case (not dashed) */
   if (orientation == XmHORIZONTAL) {
       ndash = ((width - 2*margin)/shadow_dash_size + 1)/2 ;
       for (i=0; i<ndash; i++)
           if (shadow_thick < 4) {
	       XDrawLine(display, d, top_gc,
			 x + margin + 2*i*shadow_dash_size,
			 center - shadow_thick/2,
			 x + margin + (2*i + 1)*shadow_dash_size -1,
			 center - shadow_thick/2);
	       if (shadow_thick > 1)
		   XDrawLine(display, d, bottom_gc,
			 x + margin + 2*i*shadow_dash_size,
			 center,
			 x + margin + (2*i + 1)*shadow_dash_size -1,
			 center);
	   } else {
	       DrawSimpleShadow(display, d, top_gc, bottom_gc,
				  x + margin + i*2*shadow_dash_size,
				  center - shadow_thick/2,
				  shadow_dash_size, (shadow_thick/2)*2,
				  shadow_thick/2, 0);
	   }
       /* draw the last dash, with possibly a different size */
       if (i*2*shadow_dash_size < (width - 2*margin))
       {
           if (shadow_thick < 4)
           {
	       XDrawLine(display, d, top_gc,
			 x + margin + 2*i*shadow_dash_size,
			 center - shadow_thick/2,
			 x + (width - 2*margin),
			 center - shadow_thick/2);
	       if (shadow_thick > 1)
               {
		   XDrawLine(display, d, bottom_gc,
			 x + margin + 2*i*shadow_dash_size,
			 center,
			 x + (width - 2*margin),
			 center);
               }
	   } else {
	       DrawSimpleShadow(display, d, top_gc, bottom_gc,
				  x + margin + i*2*shadow_dash_size,
				  center - shadow_thick/2,
				  (width - 2*margin) - i*2*shadow_dash_size,
				  (shadow_thick/2)*2,
				  shadow_thick/2, 0);
	   }
       }
   } else {
       ndash = ((height - 2*margin)/shadow_dash_size + 1)/2 ;
       for (i=0; i<ndash; i++)
           if (shadow_thick < 4) {
	       XDrawLine(display, d, top_gc,
			 center - shadow_thick/2,
			 y + margin + 2*i*shadow_dash_size,
			 center - shadow_thick/2,
			 y + margin + (2*i + 1)*shadow_dash_size -1);
	       if (shadow_thick > 1)
		   XDrawLine(display, d, bottom_gc,
			 center,
			 y + margin + 2*i*shadow_dash_size,
			 center,
			 y + margin + (2*i + 1)*shadow_dash_size -1);
	   } else {
	       DrawSimpleShadow(display, d, top_gc, bottom_gc,
				  center - shadow_thick/2,
				  y + margin + i*2*shadow_dash_size,
				  (shadow_thick/2)*2, shadow_dash_size,
				  shadow_thick/2, 0);
	   }
       if (i*2*shadow_dash_size < (height - 2*margin))
       {
           if (shadow_thick < 4)
           {
	       XDrawLine(display, d, top_gc,
			 center - shadow_thick/2,
			 y + margin + 2*i*shadow_dash_size,
			 center - shadow_thick/2,
			 y + (height - 2*margin));
	       if (shadow_thick > 1)
		   XDrawLine(display, d, bottom_gc,
			 center,
			 y + margin + 2*i*shadow_dash_size,
			 center,
			 y + (height - 2*margin));
	   } else {
	       DrawSimpleShadow(display, d, top_gc, bottom_gc,
				  center - shadow_thick/2,
				  y + margin + i*2*shadow_dash_size,
				  (shadow_thick/2)*2,
				  (height - 2*margin) - i*2*shadow_dash_size,
				  shadow_thick/2, 0);
           }
        }
   }
   _XmAppUnlock(app);
}

